src/legacy/reduce.rs

```diff
@@ -0,0 +1,297 @@
+use std::collections::VecDeque;
+use std::io::{self, Read, Result};
+
+use crate::legacy::lz77::lz77_output_backref;
+use bitstream_io::{BitRead, BitReader, Endianness, LittleEndian};
+
+/// Number of bits used to represent indices in a follower set of size n.
+fn follower_idx_bw(n: u8) -> u8 {
+    debug_assert!(n <= 32);
+    match n {
+        0 => 0,
+        1 => 1,
+        _ => 8 - (n - 1).leading_zeros() as u8,
+    }
+}
+
+#[derive(Default, Clone, Copy)]
+struct FollowerSet {
+    size: u8,
+    idx_bw: u8,
+    followers: [u8; 32],
+}
+
+/// Read the follower sets from is into fsets. Returns true on success.
+type FollowerSetArray = [FollowerSet; u8::MAX as usize + 1];
+
+fn read_follower_sets<T: std::io::Read, E: Endianness>(
+    is: &mut BitReader<T, E>,
+) -> io::Result<FollowerSetArray> {
+    let mut fsets = [FollowerSet::default(); u8::MAX as usize + 1];
+    for i in (0..=u8::MAX as usize).rev() {
+        let n = is.read::<u8>(6)?;
+        if n > 32 {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "invalid follower set",
+            ));
+        }
+        fsets[i].size = n;
+        fsets[i].idx_bw = follower_idx_bw(n);
+
+        for j in 0..fsets[i].size as usize {
+            fsets[i].followers[j] = is.read::<u8>(8)?;
+        }
+    }
+
+    Ok(fsets)
+}
+
+
+/// Read the next byte from is, decoded based on prev_byte and the follower sets.
+/// The byte is returned in *out_byte. 
+/// 
+/// # Returns
+/// 
+/// * `Ok(())` if the byte was successfully read.
+/// * `Err(io::Error)` on bad data or end of input.
+fn read_next_byte<T: std::io::Read, E: Endianness>(
+    is: &mut BitReader<T, E>,
+    prev_byte: u8,
+    fsets: &mut FollowerSetArray,
+) -> io::Result<u8> {
+    if fsets[prev_byte as usize].size == 0 // No followers
+            || is.read::<u8>(1)? == 1
+    // Indicates next symbol is a literal byte
+    {
+        return Ok(is.read::<u8>(8)?);
+    }
+
+    // The bits represent the index of a follower byte.
+    let idx_bw = fsets[prev_byte as usize].idx_bw;
+    let follower_idx = is.read::<u16>(idx_bw as u32)? as usize;
+    if follower_idx >= fsets[prev_byte as usize].size as usize {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "invalid follower index",
+        ));
+    }
+    Ok(fsets[prev_byte as usize].followers[follower_idx])
+}
+
+fn max_len(comp_factor: u8) -> usize {
+    let v_len_bits = (8 - comp_factor) as usize;
+    debug_assert!(comp_factor >= 1 && comp_factor <= 4);
```

Comment:
Move this up a line, since it's more useful than the overflow error that will occur on the previous line if comp_factor > 8.